### PR TITLE
[Fix] Resolve GPU Memory Leak in update_weights_from_tensor

### PR DESCRIPTION
**Pull request by U-rara**
## Motivation

My mentor @fukun07 and I discovered a significant CUDA memory leak in the `update_weights_from_tensor` functionality when tensor parallelism (TP) > 1. The input CUDA tensors used for updates fail to properly release their GPU memory after their lifecycle completes.

The root cause relates to how PyTorch handles CUDA IPC tensor sharing across processes:

1. When a CUDA tensor is serialized, PyTorch creates a new reference counter (see [StorageSharing.cpp#L331](https://github.com/pytorch/pytorch/blob/740ce0fa5f8c7e9e51422b614f8187ab93a60b8b/torch/csrc/StorageSharing.cpp#L331)) to manage the tensor's reference count across processes.

2. During deserialization, a `release_ipc_counter` operation occurs (see [reductions.py#L196](https://github.com/pytorch/pytorch/blob/740ce0fa5f8c7e9e51422b614f8187ab93a60b8b/torch/multiprocessing/reductions.py#L196)).

3. In normal usage, one serialization and one deserialization will bring the CUDA tensor's IPC reference count to zero, allowing memory to be reclaimed.

However, with TP > 1, `update_weights_from_tensor` performs one serialization but multiple deserializations (one per TP worker). With the current PyTorch implementation, this imbalance prevents the tensor from being properly garbage collected, causing a permanent memory leak.

## Modifications

To fix this issue, we need to balance the serialization and deserialization operations by performing `tp_size` serializations and one deserialization per worker. This creates separate reference counters that all point to the same CUDA tensor without additional memory overhead.

Our solution:

1. Modify the Engine to create `tp_size` serializations when generating the update request
2. Have each TP worker consume only its corresponding serialized tensor based on its `tp_rank`

```python
# Before: One serialization, multiple deserializations
obj = UpdateWeightsFromTensorReqInput(
    serialized_named_tensors=MultiprocessingSerializer.serialize(named_tensors),
    load_format=load_format,
    flush_cache=flush_cache,
)

# After: Multiple serializations, one deserialization per worker
obj = UpdateWeightsFromTensorReqInput(
    serialized_named_tensors=[
        MultiprocessingSerializer.serialize(named_tensors) for _ in
        range(self.server_args.tp_size)
    ],
    load_format=load_format,
    flush_cache=flush_cache,
)
```
And in the TP worker:
```python
# Before: All workers deserialize the same serialized tensor
success, message = self.model_runner.update_weights_from_tensor(
    named_tensors=MultiprocessingSerializer.deserialize(
        recv_req.serialized_named_tensors
    ),
    load_format=recv_req.load_format,
)

# After: Each worker deserializes its corresponding serialized tensor
success, message = self.model_runner.update_weights_from_tensor(
    named_tensors=MultiprocessingSerializer.deserialize(
        recv_req.serialized_named_tensors[self.tp_rank]
    ),
    load_format=recv_req.load_format,
)
```

I've verified this approach creates `tp_size` independent reference counters pointing to the same CUDA tensor. A simple test demonstrates this:
```shell
>>> import torch
>>> a = torch.randn(5,5).cuda()
>>> a.untyped_storage()._share_cuda_()
(0, b'\x01c\xa0-]\x06\x00\x00\x00\x00\xf2 \x00\x00\x00\x00\x00\x00\x00\x00 \x00\x00\x00\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00\x04\x00\x00\x00\x00\x00\x00,\x00\x00\x00\x00\x00\x00\x00u,\xd0\xc1\xb4\x00\x00\\\x00\x00\x00\x00\x00\x00\x00\x00', 100, 0, b'/torch_8434_2846715995_0', 0, b'\xf2 \x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\xff\x0f\x00\x00\x00\x00\x00\x00y\x03=\xe3q\x7f\x00\x00h\xbco\x90q\x7f\x00\x00\x08\t\xf8\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xc0\x08\xf8\x00\x00\x00\x00\x00', True)
>>> a.untyped_storage()._share_cuda_()
(0, b'\x01c\xa0-]\x06\x00\x00\x00\x00\xf2 \x00\x00\x00\x00\x00\x00\x00\x00 \x00\x00\x00\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00\x04\x00\x00\x00\x00\x00\x00,\x00\x00\x00\x00\x00\x00\x00u,\xd0\xc1\xb4\x00\x00\\\x00\x00\x00\x00\x00\x00\x00\x00', 100, 0, b'/torch_8434_2846715995_0', 1, b'\xf2 \x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\xfe\x0f\x00\x00\x00\x00\x00\x00y\x03=\xe3q\x7f\x00\x00h\xbco\x90q\x7f\x00\x00\x08\t\xf8\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\xc0\x08\xf8\x00\x00\x00\x00\x00', True)
>>> a.untyped_storage()._share_cuda_()
(0, b'\x01c\xa0-]\x06\x00\x00\x00\x00\xf2 \x00\x00\x00\x00\x00\x00\x00\x00 \x00\x00\x00\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00\x04\x00\x00\x00\x00\x00\x00,\x00\x00\x00\x00\x00\x00\x00u,\xd0\xc1\xb4\x00\x00\\\x00\x00\x00\x00\x00\x00\x00\x00', 100, 0, b'/torch_8434_2846715995_0', 2, b'\xf2 \x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\xfd\x0f\x00\x00\x00\x00\x00\x00y\x03=\xe3q\x7f\x00\x00h\xbco\x90q\x7f\x00\x00\x08\t\xf8\x00\x00\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00\xc0\x08\xf8\x00\x00\x00\x00\x00', True)
>>> a.untyped_storage()._share_cuda_()
(0, b'\x01c\xa0-]\x06\x00\x00\x00\x00\xf2 \x00\x00\x00\x00\x00\x00\x00\x00 \x00\x00\x00\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00\x04\x00\x00\x00\x00\x00\x00,\x00\x00\x00\x00\x00\x00\x00u,\xd0\xc1\xb4\x00\x00\\\x00\x00\x00\x00\x00\x00\x00\x00', 100, 0, b'/torch_8434_2846715995_0', 3, b'\xf2 \x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\xfc\x0f\x00\x00\x00\x00\x00\x00y\x03=\xe3q\x7f\x00\x00h\xbco\x90q\x7f\x00\x00\x08\t\xf8\x00\x00\x00\x00\x00\x03\x00\x00\x00\x00\x00\x00\x00\xc0\x08\xf8\x00\x00\x00\x00\x00', True)
```
Note how the reference counter offset (5th element in each tuple) increments from 0 to 3, indicating separate reference counters for the same tensor.

I've enhanced the previous unit tests to include CUDA tensor update testing with memory leak detection and added TP=2 test cases to ensure the fix works properly.

## Checklist

- [X] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [X] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by merrymercy** [comment]
@fzyzcjy  Can you take a look?

File: test/srt/test_update_weights_from_tensor.py [hunk 1]
```diff
@@ -9,13 +10,48 @@
...
+        del new_tensor
+        gc.collect()
+        torch.cuda.ipc_collect()
+        torch.cuda.empty_cache()
+        memory_after = torch.cuda.memory_allocated()
+        assert (
+            memory_after <= memory_before + 1024
+        ), f"Memory leak detected: {memory_after - memory_before} bytes"
+
+    def test_update_weights_from_tensor_tp2(self):
```
Comment (fzyzcjy) [inline test/srt/test_update_weights_from_tensor.py]: LGTM, only a nit: Maybe we can combine the `test_update_weights_from_tensor` and `test_update_weights_from_tensor_tp2` into one parameterized test to slightly remove code duplication.

Comment (U-rara) [reply test/srt/test_update_weights_from_tensor.py]: Great suggestion, I've completed the changes.

**Comment by fzyzcjy** [comment]
LGTM, I have no more nits, and great fix!

**Review by merrymercy — APPROVED** [review]

**Comment by xiezhq-hermann** [comment]
Nice catch! Could you help me understand why the original implementation causes a memory leak instead of just referencing null pointers? If there’s some undefined behavior with the garbage collector, wouldn’t it be better to manually release the memory after the TP sync, rather than duplicating the tensors and hoping the GC behaves correctly?

**Comment by U-rara** [comment]
Great question. To try to answer your inquiry, we studied the CUDA IPC related implementations (including [StorageSharing.cpp](https://github.com/pytorch/pytorch/blob/916e8979d3e0d651a9091732ce3e59da32e72b0e/torch/csrc/StorageSharing.cpp), [reductions.py](https://github.com/pytorch/pytorch/blob/916e8979d3e0d651a9091732ce3e59da32e72b0e/torch/multiprocessing/reductions.py), and [cuda_multiprocessing.md](https://github.com/pytorch/pytorch/blob/916e8979d3e0d651a9091732ce3e59da32e72b0e/torch/multiprocessing/cuda_multiprocessing.md)), but due to our unfamiliarity with PyTorch's underlying logic, we haven't been able to determine with 100% certainty what causes the "original implementation causes a memory leak instead of just referencing null pointers" that you mentioned.

We suspect this is due to PyTorch's reference counting management logic for IPC, which prevents the program termination issues from occurring (we need someone familiar with PyTorch or CUDA, like you, to help confirm the root cause 🥹. However, our experiments can confirm that this leak exists, and our fix is a good solution).

Regarding "release the memory after the TP sync" it's a good idea and was what we initially attempted to do, but we quickly realized it wasn't possible. We couldn't use PyTorch's API to do anything like clearing certain reference counts or forcing the release of a specific GPU memory block. So we used the approach that aligns with PyTorch's expected way of sharing CUDA tensors. It's worth noting that we're not "duplicating the tensors" but instead allocating a new reference counter for each TP worker, avoiding the aforementioned errors.

By the way, we discovered in our experiments that using:
```python
  def update_weights_from_tensor(self, recv_req: UpdateWeightsFromTensorReqInput):
    deserialized = MultiprocessingSerializer.deserialize(
        recv_req.serialized_named_tensors
    )
    success, message = self.model_runner.update_weights_from_tensor(
        named_tensors=deserialized,
        load_format=recv_req.load_format,
    )
    sleep(1 * self.tp_rank)
    del deserialized
    import gc
    gc.collect()
    torch.cuda.ipc_collect()
    torch.cuda.empty_cache()
    return success, message
```
This approach of having different TP workers release references to deserialized tensors at different times mysteriously made the memory leak disappear. However, when using `sleep(0.1 * self.tp_rank)`, the memory leak still persisted, which puzzled us greatly. Perhaps our fix completely avoided something like a race condition that was causing the memory leak.

**Comment by xiezhq-hermann** [comment]
> Great question. To try to answer your inquiry, we studied the CUDA IPC related implementations (including [StorageSharing.cpp](https://github.com/pytorch/pytorch/blob/916e8979d3e0d651a9091732ce3e59da32e72b0e/torch/csrc/StorageSharing.cpp), [reductions.py](https://github.com/pytorch/pytorch/blob/916e8979d3e0d651a9091732ce3e59da32e72b0e/torch/multiprocessing/reductions.py), and [cuda_multiprocessing.md](https://github.com/pytorch/pytorch/blob/916e8979d3e0d651a9091732ce3e59da32e72b0e/torch/multiprocessing/cuda_multiprocessing.md)), but due to our unfamiliarity with PyTorch's underlying logic, we haven't been able to determine with 100% certainty what causes the "original implementation causes a memory leak instead of just referencing null pointers" that you mentioned.
> 
> We suspect this is due to PyTorch's reference counting management logic for IPC, which prevents the program termination issues from occurring (we need someone familiar with PyTorch or CUDA, like you, to help confirm the root cause 🥹. However, our experiments can confirm that this leak exists, and our fix is a good solution).
> 
> Regarding "release the memory after the TP sync" it's a good idea and was what we initially attempted to do, but we quickly realized it wasn't possible. We couldn't use PyTorch's API to do anything like clearing certain reference counts or forcing the release of a specific GPU memory block. So we used the approach that aligns with PyTorch's expected way of sharing CUDA tensors. It's worth noting that we're not "duplicating the tensors" but instead allocating a new reference counter for each TP worker, avoiding the aforementioned errors.
> 
> By the way, we discovered in our experiments that using:
> 
> ```python
>   def update_weights_from_tensor(self, recv_req: UpdateWeightsFromTensorReqInput):
>     deserialized = MultiprocessingSerializer.deserialize(
>         recv_req.serialized_named_tensors
>     )
>     success, message = self.model_runner.update_weights_from_tensor(
>         named_tensors=deserialized,
>         load_format=recv_req.load_format,
>     )
>     sleep(1 * self.tp_rank)
>     del deserialized
>     import gc
>     gc.collect()
>     torch.cuda.ipc_collect()
>     torch.cuda.empty_cache()
>     return success, message
> ```
> 
> This approach of having different TP workers release references to deserialized tensors at different times mysteriously made the memory leak disappear. However, when using `sleep(0.1 * self.tp_rank)`, the memory leak still persisted, which puzzled us greatly. Perhaps our fix completely avoided something like a race condition that was causing the memory leak.

Thank you for the detailed explanation, I suspect it might have something to do with the implementation of MultiprocessingSerializer and reference counting across processes itself seems tricky already.
I approve this PR, but if you are interested, would you mind trying something like this:
```
# on each tp worker
del deserialized
# on rank=0 worker or all tp workers
recv_req.serialized_named_tensors = None
```

**Comment by U-rara** [comment]
> > Great question. To try to answer your inquiry, we studied the CUDA IPC related implementations (including [StorageSharing.cpp](https://github.com/pytorch/pytorch/blob/916e8979d3e0d651a9091732ce3e59da32e72b0e/torch/csrc/StorageSharing.cpp), [reductions.py](https://github.com/pytorch/pytorch/blob/916e8979d3e0d651a9091732ce3e59da32e72b0e/torch/multiprocessing/reductions.py), and [cuda_multiprocessing.md](https://github.com/pytorch/pytorch/blob/916e8979d3e0d651a9091732ce3e59da32e72b0e/torch/multiprocessing/cuda_multiprocessing.md)), but due to our unfamiliarity with PyTorch's underlying logic, we haven't been able to determine with 100% certainty what causes the "original implementation causes a memory leak instead of just referencing null pointers" that you mentioned.
> > We suspect this is due to PyTorch's reference counting management logic for IPC, which prevents the program termination issues from occurring (we need someone familiar with PyTorch or CUDA, like you, to help confirm the root cause 🥹. However, our experiments can confirm that this leak exists, and our fix is a good solution).
> > Regarding "release the memory after the TP sync" it's a good idea and was what we initially attempted to do, but we quickly realized it wasn't possible. We couldn't use PyTorch's API to do anything like clearing certain reference counts or forcing the release of a specific GPU memory block. So we used the approach that aligns with PyTorch's expected way of sharing CUDA tensors. It's worth noting that we're not "duplicating the tensors" but instead allocating a new reference counter for each TP worker, avoiding the aforementioned errors.
> > By the way, we discovered in our experiments that using:
> > ```python
> >   def update_weights_from_tensor(self, recv_req: UpdateWeightsFromTensorReqInput):
> >     deserialized = MultiprocessingSerializer.deserialize(
> >         recv_req.serialized_named_tensors
> >     )
> >     success, message = self.model_runner.update_weights_from_tensor(
> >         named_tensors=deserialized,
> >         load_format=recv_req.load_format,
> >     )
> >     sleep(1 * self.tp_rank)
> >     del deserialized
> >     import gc
> >     gc.collect()
> >     torch.cuda.ipc_collect()
> >     torch.cuda.empty_cache()
> >     return success, message
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This approach of having different TP workers release references to deserialized tensors at different times mysteriously made the memory leak disappear. However, when using `sleep(0.1 * self.tp_rank)`, the memory leak still persisted, which puzzled us greatly. Perhaps our fix completely avoided something like a race condition that was causing the memory leak.
> 
> Thank you for the detailed explanation, I suspect it might have something to do with the implementation of MultiprocessingSerializer and reference counting across processes itself seems tricky already. I approve this PR, but if you are interested, would you mind trying something like this:
> 
> ```
> # on each tp worker
> del deserialized
> # on rank=0 worker or all tp workers
> recv_req.serialized_named_tensors = None
> ```

I just tried it, but the memory leak still exists. This issue might still be related to PyTorch itself. The other related code in sglang doesn't seem to cause additional references😋.
